src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -0,0 +1,1020 @@
+                {
+                    baseData = s_baseDataMap.GetValue(data, _data => new ReferenceMap(_data));
+                }
+
+                return baseData;
+            }
+
+            public int GetNextReferenceId()
+            {
+                _values.Add(null);
+                return _baseDataCount + _values.Count - 1;
+            }
+
+            public void AddValue(int referenceId, object value)
+            {
+                _values[referenceId - _baseDataCount] = value;
+            }
+
+            public object GetValue(int referenceId)
+            {
+                if (_baseData != null)
+                {
+                    if (referenceId < _baseDataCount)
+                    {
+                        return _baseData.GetValue(referenceId);
+                    }
+                    else
+                    {
+                        return _values[referenceId - _baseDataCount];
+                    }
+                }
+                else
+                {
+                    return _values[referenceId];
+                }
+            }
+        }
+
+        internal uint ReadCompressedUInt()
+        {
+            var info = _reader.ReadByte();
+            byte marker = (byte)(info & StreamObjectWriter.ByteMarkerMask);
+            byte byte0 = (byte)(info & ~StreamObjectWriter.ByteMarkerMask);
+
+            if (marker == StreamObjectWriter.Byte1Marker)
+            {
+                return byte0;
+            }
+
+            if (marker == StreamObjectWriter.Byte2Marker)
+            {
+                var byte1 = _reader.ReadByte();
+                return (((uint)byte0) << 8) | byte1;
+            }
+
+            if (marker == StreamObjectWriter.Byte4Marker)
+            {
+                var byte1 = _reader.ReadByte();
+                var byte2 = _reader.ReadByte();
+                var byte3 = _reader.ReadByte();
+
+                return (((uint)byte0) << 24) | (((uint)byte1) << 16) | (((uint)byte2) << 8) | byte3;
+            }
+
+            throw ExceptionUtilities.UnexpectedValue(marker);
+        }
+
+        private string ReadStringValue()
+        {
+            var kind = (EncodingKind)_reader.ReadByte();
+            return kind == EncodingKind.Null ? null : ReadStringValue(kind);
+        }
+
+        private string ReadStringValue(EncodingKind kind)
+        {
+            switch (kind)
+            {
+                case EncodingKind.StringRef_B:
+                    return (string)_referenceMap.GetValue(_reader.ReadByte());
+
+                case EncodingKind.StringRef_S:
+                    return (string)_referenceMap.GetValue(_reader.ReadUInt16());
+
+                case EncodingKind.StringRef:
+                    return (string)_referenceMap.GetValue(_reader.ReadInt32());
+
+                case EncodingKind.StringUtf16:
+                case EncodingKind.StringUtf8:
+                    return ReadStringLiteral(kind);
```

Comment:
These comments are *really* good to put in the code :)    Basically, consider any any question that is #ByDesigned as something we should have comments for so that the reason is clear 1+ year down the line when someone is maintaining htis.